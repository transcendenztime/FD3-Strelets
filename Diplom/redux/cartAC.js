const QTY_ADD = 'QTY_ADD';
const PROD_ADD = 'PROD_ADD';
const PROD_DEL = 'PROD_DEL';
const SUM_QTY = 'SUM_QTY';
const QTY_TO_NULL = 'QTY_TO_NULL';
const DELIVERY_CHNG = 'DELIVERY_CHNG';

const CART_CLEAR = 'CART_CLEAR';

const prod_add = function(prodid, info) {
  return {
    type: PROD_ADD,
    prodid: prodid,
    info: info,
  };
}

const prod_del = function(prodid) {
  return {
    type: PROD_DEL,
    prodid: prodid,
  };
}

const cart_clear = function() {
  return {
    type: CART_CLEAR,
  };
}

const qty_add = function(prodid, newQTY, newSum) {
  return {
    type: QTY_ADD,
    prodid: prodid,
    newQTY: newQTY,
    newSum: newSum
  };
}

const sum_qty = function(qty) {
  return {
    type: SUM_QTY,
    qty: qty,
  }
}

const qty_to_null = function() {
  return {
    type: QTY_TO_NULL,
  }
}

const delivery_chng = function(num) {
  return {
    type: DELIVERY_CHNG,
    num: num,
  }
}

export {
  prod_add, PROD_ADD,
  prod_del, PROD_DEL,
  cart_clear, CART_CLEAR,
  qty_add, QTY_ADD,
  sum_qty, SUM_QTY,
  qty_to_null, QTY_TO_NULL,
  delivery_chng, DELIVERY_CHNG,
}
