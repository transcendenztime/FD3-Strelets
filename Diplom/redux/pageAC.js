const PAGE_CHANGE = 'PAGE_CHANGE';

const page_change = function(page) {
  return {
    type: PAGE_CHANGE,
    page: page,
  };
}

export {
  page_change, PAGE_CHANGE,
}
