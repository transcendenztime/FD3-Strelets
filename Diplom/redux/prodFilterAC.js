const PROD_FILTER = 'PROD_FILTER';

const prod_filter = function(name) {
  return {
    type: PROD_FILTER,
    name,
  };
}

export {
    prod_filter, PROD_FILTER,
}