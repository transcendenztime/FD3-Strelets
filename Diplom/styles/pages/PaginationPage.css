.pagination {    
    list-style-type: none;
    padding: 0;
    margin: 0;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    height: 50px;
}
.pagination-item {
    margin: .5px;
    display: block;
}

.pagination-item a {
    display: block;
    margin: 0;
    padding: 0;
    width: 25px;
    height: 25px;
    line-height: 25px;
    text-align: center;
    text-decoration: none;
    color: #000;
    font-weight: normal;
    border: 1px solid #dbdbdb;
    border-radius: 4px;
    background: #dbdbdb;
    background: -webkit-linear-gradient(right, #dbdbdb 0%, #fefefe 80%);
    background: -webkit-gradient(linear, left top, right top, color-stop(0%, #dbdbdb), color-stop(80%, #fefefe));
    background: -moz-linear-gradient(right, #dbdbdb 0%, #fefefe 80%);
    background: -ms-linear-gradient(right, #dbdbdb 0%, #fefefe 80%);
    background: -o-linear-gradient(right, #dbdbdb 0%, #fefefe 80%);
    background: linear-gradient(to left, #dbdbdb 0%, #fefefe 80%);
}

.pagination-item a:hover {
    color: #0d897f;
    font-weight: bold;
}
.pagination-item a:active {
    color: #0d897f;
    font-weight: bold;
    text-shadow: 1px 1px 2px black;
    box-shadow: 1px 1px 2px black;
}
.pagination-item a.activePage {
    color: #0d897f;
    font-weight: bold; 
}