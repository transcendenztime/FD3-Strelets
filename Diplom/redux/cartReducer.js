import {  PROD_ADD, PROD_DEL, CART_CLEAR, QTY_ADD, DELIVERY_CHNG } from './cartAC';

const initState={

  // ключ - идентификатор prod, значение - info
  products: {},
  delivery: 0,  
}


function cartReducer(state = initState, action) {
  switch (action.type) {
    
    case PROD_ADD: {
      let newProdInfo = action.info;
      
      //если в корзине уже есть данный товар, суммируем добавляемое с тем, что есть
      if (state.products[action.prodid]) {
        newProdInfo.qty += state.products[action.prodid].qty;
        newProdInfo.sum += state.products[action.prodid].sum;
      }

      let newState = {...state,
        products: {...state.products,
          [action.prodid]: newProdInfo}};
      return newState;
    }    

    case PROD_DEL: {
      let newState = {...state};
      delete newState.products[action.prodid];
      return newState;
    }    

    case CART_CLEAR: {
      //чистим корзину
      let newState = {...initState};
      return newState;
    }

    case QTY_ADD: {
      //меняем количество единиц товара и общую сумму денег за это количество единиц товара 
      let newState = {...state,
        products: {...state.products,
          [action.prodid]: {...state.products[action.prodid],
            qty: action.newQTY, sum: action.newSum }
        }
      };

      return newState;
    }
    
    case DELIVERY_CHNG: {
      let newState = {...state,
        delivery: action.num};
      return newState;
    } 
    
    default:
      return state;
  }
}

export default cartReducer;
