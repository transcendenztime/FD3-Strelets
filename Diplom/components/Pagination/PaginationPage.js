import React from 'react';
import PropTypes from 'prop-types';

import {connect} from 'react-redux';
import { page_change } from "../../redux/pageAC";

import PaginationItem from './PaginationItem';
import Product from '../Catalogue/Product';

import '../../styles/pages/PaginationPage.css';

class PaginationPage extends React.PureComponent {
         
  static propTypes = {
    products: PropTypes.array.isRequired, //передан из родительского компонента
    startLink: PropTypes.string.isRequired, //передан из родительского компонента
    page: PropTypes.object.isRequired,
    pagesQty: PropTypes.number.isRequired,//передан из родительского компонента разбивать ли на страницы
  };

  componentWillUnmount() {
    this.props.dispatch( page_change(1) );
  }
  
  render() {
    
    let prodData;
    let pageArr = [];

    if(this.props.pagesQty == 1) {
      prodData = [];
      let start = (this.props.page.page - 1) * 25; //по 25 продукта на странице
      let end = this.props.page.page * 25;
      for(let i = start; i < end; i++ ) {
          if(this.props.products[i])
            prodData.push(this.props.products[i]);
      }   
    
      let l = this.props.products.length;

      pageArr = [];
      let pageQty = Math.ceil(l / 25);//по 25 продукта на странице
      for(let i = 1; i <= pageQty; i++) {        
          pageArr.push(<PaginationItem key = {i} num = {i} startLink = {this.props.startLink} />)
        }
    } else {
      prodData = this.props.products
    }


    let productsArr = [];
    for(let i=0; i < prodData.length; i++) {
      let prod = prodData[i];
      productsArr.push(<Product key = {prod.id} info = {prod} page = {this.props.page.page} />)
    }

    return(
        <div className = "catalogue page">
        {pageArr.length > 1? 
          <ul className = "pagination">
              {pageArr}
          </ul>
          :null
        }
          <div className = "products-container">
            {productsArr}
          </div>        
            
        </div>
    );
  }

}
      
const mapStateToProps = function (state) {
  return {    
    page: state.page,
  };
};

export default connect(mapStateToProps)(PaginationPage);