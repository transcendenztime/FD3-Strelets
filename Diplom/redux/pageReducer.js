import {  PAGE_CHANGE } from './pageAC';

const initState={
    page: 1,
}

function pageReducer(state = initState, action) {
  switch (action.type) {
    
    case PAGE_CHANGE: {
      
      //если страница та же, ничего не делаем
      if (state.page != action.page) {
        let newState = {...state,
          page: action.page};
        return newState;
      } else {return state}
      
    }    
    
    default:
      return state;
  }
}

export default pageReducer;
