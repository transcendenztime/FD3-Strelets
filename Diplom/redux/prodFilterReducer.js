import {  PROD_FILTER } from './prodFilterAC';

const initState={
    prodFilter: '',
}


function prodFilterReducer(state = initState, action) {
  switch (action.type) {
    
    case PROD_FILTER: {
            
      let newState = {...state,
        prodFilter: action.name};
        console.log(action.name)
      return newState;
      
    }    
    
    default:
      return state;
  }
}

export default prodFilterReducer;
