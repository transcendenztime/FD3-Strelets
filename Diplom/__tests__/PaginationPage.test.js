"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import { createStore } from 'redux';
import combinedReducer from '../redux/reducers.js';
import { Provider } from 'react-redux';

import PaginationPage from '../components/Pagination/PaginationPage';
  
const prods = [{
    "id": 1,
    "name": "Томат Гранатовая капля",
    "brand": "Гавриш",
    "category": "Семена овощей",
    "price": 1.32,
    "description": "Сладкий фруктовый вкус. Среднеспелый сорт для пленочных теплиц и открытого грунта.",
    "imgUrl": "./img/ed940bf6db9540b843b64bacabbae4ae.jpg",
    "qty": 1,
    "sum": 1.32
  },
  {
    "id": 2,
    "name": "Свекла Смуглянка",
    "brand": "Гавриш",
    "category": "Семена овощей",
    "price": 0.99,
    "description": "Среднеспелый сорт столовой свёклы с превосходными вкусовыми качествами.",
    "imgUrl": "./img/f19a9bbfa325e1ae0b3e6bd185602826.jpg",
    "qty": 1,
    "sum": 0.99
  }
];

let store=createStore(combinedReducer);

test('внешний вид и работа PaginationPage', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <Router>
      <Provider store={store}>
        <PaginationPage products = {prods}  startLink = {'/catalogue'} pagesQty = {0}/>
      </Provider>
    </Router>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // создаём 2 тестовую версию компонента
  const component2 = renderer.create(
    <Router>
      <Provider store={store}>
        <PaginationPage products = {prods}  startLink = {'/catalogue/category'} pagesQty = {1}/>
      </Provider>
    </Router>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree2=component2.toJSON();
  expect(componentTree2).toMatchSnapshot();
        
});
