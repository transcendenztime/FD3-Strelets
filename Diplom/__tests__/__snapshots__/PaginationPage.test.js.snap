// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`внешний вид и работа PaginationPage 1`] = `
<div
  className="catalogue page"
>
  <div
    className="products-container"
  >
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/semena-ovoshchey/1"
          onClick={[Function]}
          style={undefined}
        >
          Томат Гранатовая капля
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/semena-ovoshchey/1"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/ed940bf6db9540b843b64bacabbae4ae.jpg"
        />
      </a>
      <p>
        1.32 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="1"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-shopping-cart"
          />
        </button>
        <div
          className="invisibleAddSpinner"
        >
          <i
            className="fa fa-spinner fa-spin fa-fw"
          />
        </div>
      </div>
    </div>
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/semena-ovoshchey/2"
          onClick={[Function]}
          style={undefined}
        >
          Свекла Смуглянка
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/semena-ovoshchey/2"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/f19a9bbfa325e1ae0b3e6bd185602826.jpg"
        />
      </a>
      <p>
        0.99 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="1"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-shopping-cart"
          />
        </button>
        <div
          className="invisibleAddSpinner"
        >
          <i
            className="fa fa-spinner fa-spin fa-fw"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`внешний вид и работа PaginationPage 2`] = `
<div
  className="catalogue page"
>
  <div
    className="products-container"
  >
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/semena-ovoshchey/1"
          onClick={[Function]}
          style={undefined}
        >
          Томат Гранатовая капля
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/semena-ovoshchey/1"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/ed940bf6db9540b843b64bacabbae4ae.jpg"
        />
      </a>
      <p>
        1.32 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="1"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-shopping-cart"
          />
        </button>
        <div
          className="invisibleAddSpinner"
        >
          <i
            className="fa fa-spinner fa-spin fa-fw"
          />
        </div>
      </div>
    </div>
    <div
      className="product-wrap"
    >
      <h3>
        <a
          aria-current={null}
          className="product-title"
          href="/catalogue/semena-ovoshchey/2"
          onClick={[Function]}
          style={undefined}
        >
          Свекла Смуглянка
        </a>
      </h3>
      <a
        aria-current={null}
        className="product-title product-img-container"
        href="/catalogue/semena-ovoshchey/2"
        onClick={[Function]}
        style={undefined}
      >
        <img
          className="product-img"
          src="./img/f19a9bbfa325e1ae0b3e6bd185602826.jpg"
        />
      </a>
      <p>
        0.99 руб.
      </p>
      <div
        className="addCart-block"
      >
        <input
          className="addQty_button"
          defaultValue={1}
          min="1"
          step="1"
          title="Кол-во"
          type="number"
        />
        <button
          className="addCart_button"
          onClick={[Function]}
          title="Добавить в корзину"
        >
          <i
            className="fas fa-shopping-cart"
          />
        </button>
        <div
          className="invisibleAddSpinner"
        >
          <i
            className="fa fa-spinner fa-spin fa-fw"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
